app/models.py

```diff
@@ -345,6 +402,38 @@ def supplier(cls):
     def framework(cls):
         return db.relationship(Framework, lazy='joined', innerjoin=True)
 
+    @declared_attr
+    def lot(cls):
+        return db.relationship(Lot, lazy='joined', innerjoin=True)
+
+    @validates('service_id')
+    def validate_service_id(self, key, value):
+        if not is_valid_service_id(value):
+            raise ValidationError("Invalid service ID value '{}'".format(value))
+
+        return value
+
+    @validates('status')
+    def validates_status(self, key, value):
+        if value not in self.STATUSES:
+            raise ValidationError("Invalid status value '{}'".format(value))
+
+        return value
+
+    @validates('data')
```

Comment:
I think the name is a bit misleading – essentially they're just setters for the attribute, so whatever value gets returned is stored. Validation is only achieved by raising exceptions (which is also something you'd do in a setter), so it's different from the common `if x.is_valid()` pattern.

I think it's better to keep the naming convention/methods from SQLAlchemy though (even if we could just use properties here) and keeping all the ".data -> JSON -> .data" within the model definitely feels like a good approach for us.